analog_digital_out.py

```diff
@@ -53,26 +93,61 @@ def main():
             print('  [', i, '] ', devices[i].product_name, ' (',
                   devices[i].unique_id, ')', sep='')    
     
-    
         # Create the DAQ device
         daq_device = DaqDevice(devices[descriptor_index])
         ao_device = daq_device.get_ao_device()
-        # ai_device = daq_device.get_ai_device()
+        ai_device = daq_device.get_ai_device()
         dio_device = daq_device.get_dio_device()
 
         # Establish a connection to the device.
         # For Ethernet devices using a connection_code other than the default value of zero, change the line below to enter the desired code.
         daq_device.connect(connection_code=0)
                 
         # Create an array for output data. Analog Out
-        out_buffer = create_float_buffer(number_of_channels = 1, samples_per_channel = 3000)
+        out_buffer = create_float_buffer(number_of_channels, ao_samples_per_channel)
 
         # Configure the port for output. Digital
         dio_device.d_config_port(DigitalPortType.AUXPORT, DigitalDirection.OUTPUT)
-
+       
+        # Get a list of supported ranges and validate the range index
+        ranges = ai_device.get_info().get_ranges(input_mode)
+        if range_index >= len(ranges):
+            range_index = len(ranges) - 1
+        
+        # Allocate a buffer to receive the data. Analog In
+        data = create_float_buffer(ai_channel_count, ai_samples_per_channel)
+       
+        # parameters in a_in_scan_events callback function
+        user_data = scan_params(
```

Comment:
I think `callback_params` is a better name than `user_data`

Reply:
I agree!